extensions/amp-analytics/0.1/visibility-helper.js

```diff
@@ -394,3 +425,458 @@ export class VisibilityHelper {
 function timeBase(time, baseTime) {
   return time >= baseTime ? time - baseTime : 0;
 }
+
+
+/**
+ * A base class for `VisibilityRootForDoc` and `VisibilityRootForEmbed`. The
+ * instance of this class corresponds 1:1 to `AnalyticsRoot`. It represents a
+ * collection of all visibility triggers declared within the `AnalyticsRoot`.
+ * @implements {../../../src/service.Disposable}
+ * @abstract
+ */
+export class VisibilityRoot {
+  /**
+   * @param {?VisibilityRoot} parent
+   * @param {!../../../src/service/ampdoc-impl.AmpDoc} ampdoc
+   */
+  constructor(parent, ampdoc) {
+    /** @const @protected */
+    this.parent = parent;
+
+    /** @const @protected */
+    this.ampdoc = ampdoc;
+
+    /** @const @private */
+    this.resources_ = resourcesForDoc(ampdoc);
+
+    /** @const @private {!Array<!VisibilityModel>}> */
+    this.models_ = [];
+
+    /** @const @private {!Array<!UnlistenDef>} */
+    this.unsubscribe_ = [];
+  }
+
+  /** @override */
+  dispose() {
+    // Give the chance for all events to complete.
+    this.getRootModel().setVisibility(0);
+
+    // Dispose all models.
+    this.getRootModel().dispose();
+    for (let i = this.models_.length - 1; i >= 0; i--) {
+      this.models_[i].dispose();
+    }
+
+    // Unsubscribe everything else.
+    this.unsubscribe_.forEach(unsubscribe => {
+      unsubscribe();
+    });
+    this.unsubscribe_.length = 0;
+  }
+
+  /**
+   * @param {!UnlistenDef} handler
+   */
+  unsubscribe(handler) {
+    this.unsubscribe_.push(handler);
+  }
+
+  /**
+   * The start time from which all visibility events and times are measured.
+   * @return {number}
+   * @abstract
+   */
+  getStartTime() {}
+
+  /**
+   * Whether the visibility root is currently in the background.
+   * @return {boolean}
+   * @abstract
+   */
+  isBackgrounded() {}
+
+  /**
+   * Whether the visibility root has been created in the background mode.
+   * @return {boolean}
+   * @abstract
+   */
+  isBackgroundedAtStart() {}
+
+  /**
+   * Returns the visibility model for this root. If the root is not visibile,
+   * it returns the value of 0, otherwise the value greater than 0 and less
+   * than 1.
+   * @return {!VisibilityModel}
+   * @abstract
+   */
+  getRootModel() {}
+
+  /**
+   * Listens to the visibility events on the root as the whole and the given
+   * visibility spec. The visibility tracking can be deferred until
+   * `readyPromise` is resolved, if specified.
+   * @param {!Object<string, *>} spec
+   * @param {?Promise} readySignal
+   * @param {function(!Object<string, *>)} listener
+   * @return {!UnlistenDef}
+   */
+  listenRoot(spec, readySignal, listener) {
+    const model = new VisibilityModel(
+        this.getRootModel(),
+        spec,
+        /* ownVisibility */ 1,
+        /* factorParent */ true);
```

Comment:
I see we create a model for every root spec. But why do they share a common root model? Is there any data in the root model are shared?